src/redux/actions/planActions.js

```diff
@@ -0,0 +1,329 @@
+    })
+      .then(response => response.json())
+      .then(json => {
+        if (json.length === 0) dispatch(getPlanInfoFailure());
+        else {
+          // Push newly added plan to state plan list
+          let plans = getState().masterReducer.plans.plans;
+          let trip_id = json.result[0].trip_id;
+          let e_id = json.result[0].e_id;
+          if(plans[trip_id] === undefined) plans[trip_id] = {};
+          plans[trip_id][e_id] = json.result[0];
+          dispatch(getPlanInfoSuccess(plans));
+        }
+      });
+  };
+}
+
+/**
+ * Performs an http GET plans request to server.
+ * Dispatches getPlans to indicate the beginning of a getPlans process.
+ * Dispatches getPlansFailure to indicate the end of a failed getPlans process.
+ * Dispatches getPlansSuccess to indicate the end of a successful getPlans process.
+ * If getPlans process succeeds, a plans object is received and passed into
+ * getPlansSuccess to be stored into state.
+ * @returns {function(*): Promise<Response | never>} dispatch results.
+ */
+export function getPlans() {
+  return dispatch => {
+    dispatch(getPlansRequest()); // Get plans request process has begun...
+    return fetch('http://localhost:4201/plan/get', {
+      headers: {
+        'Content-Type': 'application/json',
+        'x-access-token': localStorage.getItem('token')
+      },
+      method: 'GET'
+    })
+      .then(response => response.json())
+      .then(json => {
+        if (json.success === false) dispatch(getPlansFailure());
+        else {
+          let plans = [];
+          json.result.forEach(t => {
+            if (plans[t.trip_id] === undefined) plans[t.trip_id] = {};
+            plans[t.trip_id][t.e_id] = t;
+          });
+          dispatch(getPlansSuccess(plans));
+        }
+      });
+  };
+}
+
+/**
+ * Performs an http POST addPlan request to server.
+ * Dispatches addPlan to indicate the beginning of an addPlan process.
+ * Dispatches addPlanFailure to indicate the end of a failed addPlan process.
+ * Dispatches addPlanSuccess to indicate the end of a successful addPlan process.
+ * If addPlan process succeeds, getPlanInfo is dispatched using the returned
+ * e_id of the newly added plan.
+ * @param plan object containing email, name, and dscript.
+ * @returns {function(*): Promise<Response | never>} dispatch results.
+ */
+export function addPlan(plan) {
+  return dispatch => {
+    dispatch(addPlanRequest()); // Add plan request process has begun...
+    return fetch('http://localhost:4201/plan/add', {
+      headers: {
+        'Content-Type': 'application/json',
+        'x-access-token': localStorage.getItem('token')
+      },
+      method: 'POST',
+      body: JSON.stringify(plan)
+    })
+      .then(response => response.json())
+      .then(json => {
+        if (json.success === false) dispatch(addPlanFailure());
+        else {
+          dispatch(addPlanSuccess());
+          dispatch(getPlanInfo(json.result)); // Fetch added plan
```

Comment:
I'm not sure about this one. We know that if we had success, then the object we sent was placed into the database. Thus, can't we just directly take our local object (combined with the id) and add it to the state using a reducer? I feel doing an extra fetch is pointless when we already have all the data we need.

Reply:
I suppose that either way _would_ work, but for the sake of absolute continuity with the database I've chose to take this approach. This way we are guaranteed that whatever row is added to our database directly reflects what is inside our state, with the auto generated e_id included.

Reply:
Cool! I'm pretty neutral. Since you already did this approach, we can leave it!